### ISSUE TITLE
Angular restarts unexpectedly only in Chrome Browsers (all versions)

### ISSUE THREAD

**Issue by evh671** [angular/angular#19431]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

I have a Recipe Application which have 2 navigations: Recipes, Shopping List
(Each one of them a component).

For each component, the url should be : localhost:4200/#

The issue is that when I add an item to the shopping list array (in the "Shopping List" Component)
the URL changes unexpectedly to: 
localhost:4200/?#

That makes Angular to restart the app.

On the otherhand, using another browser such as FireFox, this issue doesn't exists if you perform 
the same scenario.
In other words, using other browsers keeps the url "localhost:4200/#" as it should be.
Hence, the app works great.

As explained above, the issue only exists in Chrome in all versions.
(I've checked each Chrome version from the last 2 years). 

## Expected behavior

Using Chrome browsers, changing the url from "localhost:4200/#" to "localhost:4200/?#"
unexpectedly shouldn't happen.

## Environment

<pre><code>
Angular version: X.Y.Z

4.3.6

Browser:
- [X] Chrome (desktop) version 61.0.3163.100 
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.3.0  
- Platform: Windows  

Others:

</code></pre>

**Comment by Toxicable** [comment]
Thanks for reporting this issue. However, you didn't provide sufficient information for us to understand and reproduce the problem. Please check out [our submission guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-submitting-an-issue) to understand why we can't act on issues that are lacking important information.

If the problem still persists, please file a new issue and ensure you provide all of the required information when filling out the issue template.

**Comment by evh671** [comment]
Dear Fabian,

Thank you for your reply.

The only way to describe it is to see it.
I've only succeeded to isolate the problem and found out that using:

 Angular v4.3.6 + NodeJS v8.3 +Chrome v61.0.3163.100,

creates unexpected behavior.

In order to give you how it happens, I'll be glad to send you a link for
the application.

I will upload it in my github and supply a link with a guidance so you
could see it.


Thanks in Advance,

Eyal Nahum

**Comment by Toxicable** [comment]
It is near impossible to identify an issue with Angular in a full Application since there are so many factors that could be causing issues.
If you want to make a reproduction then a minimal reproduction would be the best way to go about that as outlined in our submission guidelines https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-submitting-an-issue

**Comment by evh671** [comment]
I truly understand what your meaning Fabian.

Nevertheless, it's a bug that cannot be reproduced by taking a simple code.
It's more complex because it's an issue of unexpected behavior that relates
to Google Chrome or Angular's behind code.

The app works fine using FireFox.

I'll be glad if there's something else that we can promote this issue
together.
Otherwise, it could make major issues on future commercial applications.

**Comment by mlc-mlapis** [comment]
@evh671 ... be realistic ... and did you try to confirm the behavior on more devices then one ... even different Chrome versions?

**Comment by evh671** [comment]
@mlc-mlapis 
This issue is very important and the last thing I want is to spend anybody's time.
It's important that we all cooperate in order to prevent major issues to happen in the future.

I eventually, succeeded to isolate the problem and I've edited the case with all necessary details.

If there's anything else you need, I'm available.

**Comment by mlc-mlapis** [comment]
@evh671 ... good, now the problem description looks much more better. From your description it is clear that your are using `HashLocationStrategy` ... this is the `#`. Then it looks that you do some steps during adding an item to the shopping list array ... looks like navigation with query param, route param ... this is the `?`.

I have to say that I didn't see any similar case when the result is like `http://..../?#`. There should be something wrong with your code that leads to that situation. It could be also a problem in combination with only Chrome but as a result of unexpected edge case and it is certainly related to routing.

Do you have the GitHub repository of that app to try it locally?

Btw, ... do you have any extra reasons to use `HashLocationStrategy` instead of `PathLocationStrategy`?

**Comment by evh671** [comment]
Hi Miloš,
Thank you for your reply.

regarding the "HashLocationStrategy" routing, it's something I didn't add at all.
It was created by the Angular CLI when I have generated a new project.

I've uploaded the app to my github:
https://github.com/evh671/Recipe-Book

Thank you for providing me an additional help. I appreciate it.
I'm waiting for your reply.

Thanks in advance,
EVH671

**Comment by Toxicable** [comment]
I've inspected your app and I can confirm that this isn't and issue with Angular.
In your Shopping List component you have a `form` and inside that you have a button with type="submit" since you aren't using Angular's forms to bind to that event the default behavior is for it to post the data on your url, this is what causes the page to refresh. 
I would suggest walking through our tutorial on getting started with Angular  https://angular.io/tutorial which might help with things like this

**Comment by evh671** [comment]
@Toxicable 
Thank you for your reply.

I have noticed my mistake. Once I changed to "type=button" it all got fixed.
Just out of curiosity...What did you do on debugging in order to find that this is the root of the problem ?

What did guide you ?

**Comment by Toxicable** [comment]
I noticed that `?` would appear in the url after clicking add, this indicates a post, so I checked the button and as I guessed it was doing a post

**Comment by evh671** [comment]
@Toxicable @mlc-mlapis 

Great. Thank you for sharing this information with me and 
I appreciate your support.

Have a great day!
EVH671

**Comment by tbkj98** [comment]
In the ingredient adding form in **shopping list change your add button type from 'submit' to 'button'**.
Than it should be fine .
